### PR TITLE
Remove black oil props from init

### PR DESCRIPTION
**Pull request by totto82**
This PR untangle the EQUIL initialization from opm-core legacy property, blackoilState and blackoilPhase objects.

### CODE REVIEW

**Comment by bska** [comment]
> This PR untangle the EQUIL initialization

I'm not sure "untangle" is the word I'd use for this.  The word "switch" is definitely more descriptive.

The feature set looks okay, but the implementation is not.  I'm not going to merge this in its current state.  There are too many `shared_ptr<X>` parameters that should be `const X&` instead and rely on nested lifetimes.  Moreover, the "always three-phase" behaviour is a step backwards.  Why is that necessary?  Will we reenable two-phase support at a later time?  If not, why not?

Add to this a litany of white-space changes that only obscure the change-set and using [`->template`](https://github.com/OPM/opm-core/pull/1187/commits/42847d37c7b9800d7d984aac1975de1d9aecf8bd#diff-269066f88a1ce3baa7cdcedd29186fe4R105) outside of dependent names in addition to owning raw pointers that should be smart pointers.

In short, this proposal needs additional work before it is suitable for inclusion.

**Comment by totto82** [comment]
> There are too many shared_ptr<X> parameters that should be const X&

You mean the MaterialLawManager. I just followed what's the practice in the rest of the code.  

>Moreover, the "always three-phase" behaviour is a step backwards. Why is that necessary? Will we reenable two-phase support at a later time? If not, why not?

It works perfectly for 2p cases. It is just that the size of the pressure and saturations vectors are always 3 in order to use the fixed phaseIdx from opm-material. I don't think this will be very time critical anyway. 

>Add to this a litany of white-space changes that only obscure the change-set and using ->template outside of dependent names in addition to owning raw pointers that should be smart pointers

I will clean this up.

File: opm/core/simulator/EquilibrationHelpers.hpp [hunk 1]
```diff
@@ -662,70 +549,133 @@ namespace Opm
...
         private:
-            const BlackoilPropertiesInterface& props_;
+
+            std::shared_ptr<MaterialLawManager> materialLawManager_;
             const int phase_;
             const int cell_;
             const double target_pc_;
-            mutable double s_[BlackoilPhases::MaxNumPhases];
-            mutable double pc_[BlackoilPhases::MaxNumPhases];
+            mutable SatOnlyFluidState fluidState_;
```
Comment (andlaus) [inline opm/core/simulator/EquilibrationHelpers.hpp]: why does this need to mutable?

File: opm/core/simulator/EquilibrationHelpers.hpp [hunk 2]
```diff
@@ -662,70 +549,133 @@ namespace Opm
...
+            mutable double pc_[FluidSystem::numPhases];
```
Comment (andlaus) [inline opm/core/simulator/EquilibrationHelpers.hpp]: this attribute looks superfluous to me

File: opm/core/simulator/EquilibrationHelpers.hpp [hunk 3]
```diff
@@ -747,37 +697,47 @@ namespace Opm
...
-            const BlackoilPropertiesInterface& props_;
+            std::shared_ptr<MaterialLawManager> materialLawManager_;
             const int phase1_;
             const int phase2_;
             const int cell_;
             const double target_pc_;
-            mutable double s_[BlackoilPhases::MaxNumPhases];
-            mutable double pc_[BlackoilPhases::MaxNumPhases];
+            mutable SatOnlyFluidState fluidState_;
+            mutable double pc_[FluidSystem::numPhases];
```
Comment (andlaus) [inline opm/core/simulator/EquilibrationHelpers.hpp]: same as above.

File: opm/core/simulator/initStateEquil_impl.hpp [hunk 4]
```diff
@@ -176,45 +165,42 @@ namespace Opm
...
+                    double rs = rs_(depth, press, temp_);
+                    double bOil = 0.0;
+                    if ( !FluidSystem::enableDissolvedGas() || rs >= FluidSystem::oilPvt().saturatedGasDissolutionFactor(pvtRegionIdx_, temp_, press) ) {
+                        bOil = FluidSystem::oilPvt().saturatedInverseFormationVolumeFactor(pvtRegionIdx_, temp_, press);
+                    } else {
+                        bOil = FluidSystem::oilPvt().inverseFormationVolumeFactor(pvtRegionIdx_, temp_, press, rs);
+                    }
+                    double rho = bOil * FluidSystem::referenceDensity(FluidSystem::oilPhaseIdx, pvtRegionIdx_);
+                    if (FluidSystem::enableDissolvedGas()) {
+                        rho += rs * bOil * FluidSystem::referenceDensity(FluidSystem::gasPhaseIdx, pvtRegionIdx_);
```
Comment (andlaus) [inline opm/core/simulator/initStateEquil_impl.hpp]: this can be possibly simplified using `FluidSystem::density()`. That requires a fluid state where pressure, temperature and the phase compositions are already set, though.

**Comment by bska** [comment]
>> There are too many `shared_ptr<X>` parameters that should be `const X&`
>
> You mean the `MaterialLawManager`.

I do.

> I just followed what's the practice in the rest of the code.

Then this is an opportunity to improve upon the existing practice.  An interface that accepts a `shared_ptr<>` explicitly states that it wants to take part in the [lifetime management](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) of the underlying object.  The majority of this code does not need to do that.

>> Moreover, the "always three-phase" behaviour is a step backwards. Why is that necessary? Will we reenable two-phase support at a later time? If not, why not?
>
> It works perfectly for 2p cases. It is just that the size of the pressure and saturations vectors are always 3 in order to use the fixed `phaseIdx` from opm-material. I don't think this will be very time critical anyway.

Is that an internal implementation detail or does it leak out to callers?  Always carrying a three-phase saturation vector is definitely not a trivial cost.

File: opm/core/simulator/initStateEquil.hpp [hunk 5]
```diff
@@ -161,12 +142,12 @@ namespace Opm
...
          *                            one saturation value per cell in the region.
          */
-        template <class Grid, class Region, class CellRange>
+        template <class FluidSystem, class Grid, class Region, class CellRange, class MaterialLawManager>
         std::vector< std::vector<double> >
         phaseSaturations(const Grid&             grid,
                          const Region&           reg,
                          const CellRange&        cells,
-                         BlackoilPropertiesFromDeck& props,
+                         std::shared_ptr<MaterialLawManager> materialLawManager,
```
Comment (atgeirr) [inline opm/core/simulator/initStateEquil.hpp]: About the shared_ptr issue, I guess this line is typical of what @bska wants to see gone. I agree with him, this essentially says that the function needs to preserve (by sharing ownership to) the MaterialLawManager, possibly beyond the running of the function itself. It also takes a non-const object, so it can modify the MaterialLawManager, which probably is not what is needed (although I see that the old version took a non-const reference to the props object, I don't know why but if there was a good reason it could apply to the new code too).

Passing by const reference should be the default unless we require otherwise. Remember: the parser was in a similar situation, where most of the time shared_ptrs were used. Now those are only used where necessary, and usually const references are used.

Comment (totto82) [reply opm/core/simulator/initStateEquil.hpp]: This is where SWATINIT is applied so it can not be const. 
I can double-check that const is added whenever it is possible in the rest of the code . 
I can also change to passing a references instead of shared_ptr if that is preferred.

Comment (atgeirr) [reply opm/core/simulator/initStateEquil.hpp]: > This is where SWATINIT is applied so it can not be const.

Aha, thanks for the clarification. Does this have to be simultaneous though? Is it possible to first compute the saturations (with const material laws) and then *afterwards* explicitly update it for swatinit? That would make it a lot less "magical" I think.

As to const and references: yes please and yes please!

File: opm/core/simulator/initStateEquil_impl.hpp (hunk 4 above)
Comment (totto82) [reply opm/core/simulator/initStateEquil_impl.hpp]: I chose this way to avoid creating the fluid state every time I call this function, but maybe you are right. Depending on a explicit test rs >= rsSat to determine if the system is saturated or not may not be a good idea if we decide to change this logic in the future. I will try to rewrite and see how that turns out.

File: opm/core/simulator/EquilibrationHelpers.hpp (hunk 1 above)
Comment (totto82) [reply opm/core/simulator/EquilibrationHelpers.hpp]: The operator() is const

**Comment by totto82** [comment]
> Is that an internal implementation detail or does it leak out to callers? Always carrying a three-phase saturation vector is definitely not a trivial cost.

Partly. The caller needs to initializer whatever state object it uses with the correct number of phases. Both Flow and Flow-legacy only carries around the needed phases. (i.e. 2 for 2p problems) 

I changed the shared pointers to ref. as you suggested.  

> and using ->template outside of dependent names in addition to owning raw pointers that should be smart pointers.

Could you be more specific on this?

**Comment by bska** [comment]
>> and using ->template outside of dependent names in addition to owning raw pointers that should be smart pointers.
>
> Could you be more specific on this?

I mean that the `->template` syntax, or its brethren `::template` and `.template`, exist for the purpose of invoking templated identifiers (member functions, member types &c) on objects that depend on a template parameter.  They are not needed to invoke any arbitrary templated member function regardless of context and we generally should not use that syntax outside the implementation of a template (e.g., a function template).

As for owning raw pointers, your [`initDefaultFluidSystem()`](https://github.com/OPM/opm-core/pull/1187/files#diff-269066f88a1ce3baa7cdcedd29186fe4R71) function contains multiple instances of
```C++
X* p = new X(args);
```
when
```C++
auto p = make_shared<X>(args);
```
serves the same purpose in a much safer way, especially since the pointers are subsequently put into `shared_pointer<>` containers anyway.  I think you should apply the patch below and, while you're there, also declare the function `static` (or preferably put it into an anonymous namespace) to address the warnings concerning
```
"no previous declaration for ‘void initDefaultFluidSystem()’ [-Wmissing-declarations]"
```
By the way you should also remove the `const` qualifier in the [return type](https://github.com/OPM/opm-core/pull/1187/files#diff-28e715e320e8b5e39090db6aaa0fd050R541) of
```C++
const int Opm::EQUIL::EquilReg::pvtIdx() const;
```
Returning a `const int` serves no purpose.

---

```diff
diff --git a/tests/test_equil.cpp b/tests/test_equil.cpp
index 3c40a287..b41fd0da 100644
--- a/tests/test_equil.cpp
+++ b/tests/test_equil.cpp
@@ -98,25 +98,25 @@ void initDefaultFluidSystem() {
     FluidSystem::setEnableVaporizedOil(false);
     FluidSystem::setReferenceDensities(rhoRefO, rhoRefW, rhoRefG, /*regionIdx=*/0);
 
-    Opm::GasPvtMultiplexer<double> *gasPvt = new Opm::GasPvtMultiplexer<double>;
+    auto gasPvt = std::make_shared<Opm::GasPvtMultiplexer<double>>();
     gasPvt->setApproach(Opm::GasPvtMultiplexer<double>::DryGasPvt);
-    auto& dryGasPvt = gasPvt->template getRealPvt<Opm::GasPvtMultiplexer<double>::DryGasPvt>();
+    auto& dryGasPvt = gasPvt->getRealPvt<Opm::GasPvtMultiplexer<double>::DryGasPvt>();
     dryGasPvt.setNumRegions(/*numPvtRegion=*/1);
     dryGasPvt.setReferenceDensities(/*regionIdx=*/0, rhoRefO, rhoRefG, rhoRefW);
     dryGasPvt.setGasFormationVolumeFactor(/*regionIdx=*/0, Bg);
     dryGasPvt.setGasViscosity(/*regionIdx=*/0, mug);
 
-    Opm::OilPvtMultiplexer<double> *oilPvt = new Opm::OilPvtMultiplexer<double>;
+    auto oilPvt = std::make_shared<Opm::OilPvtMultiplexer<double>>();
     oilPvt->setApproach(Opm::OilPvtMultiplexer<double>::DeadOilPvt);
-    auto& deadOilPvt = oilPvt->template getRealPvt<Opm::OilPvtMultiplexer<double>::DeadOilPvt>();
+    auto& deadOilPvt = oilPvt->getRealPvt<Opm::OilPvtMultiplexer<double>::DeadOilPvt>();
     deadOilPvt.setNumRegions(/*numPvtRegion=*/1);
     deadOilPvt.setReferenceDensities(/*regionIdx=*/0, rhoRefO, rhoRefG, rhoRefW);
     deadOilPvt.setInverseOilFormationVolumeFactor(/*regionIdx=*/0, Bo);
     deadOilPvt.setOilViscosity(/*regionIdx=*/0, muo);
 
-    Opm::WaterPvtMultiplexer<double> *waterPvt = new Opm::WaterPvtMultiplexer<double>;
+    auto waterPvt = std::make_shared<Opm::WaterPvtMultiplexer<double>>();
     waterPvt->setApproach(Opm::WaterPvtMultiplexer<double>::ConstantCompressibilityWaterPvt);
-    auto& ccWaterPvt = waterPvt->template getRealPvt<Opm::WaterPvtMultiplexer<double>::ConstantCompressibilityWaterPvt>();
+    auto& ccWaterPvt = waterPvt->getRealPvt<Opm::WaterPvtMultiplexer<double>::ConstantCompressibilityWaterPvt>();
     ccWaterPvt.setNumRegions(/*numPvtRegions=*/1);
     ccWaterPvt.setReferenceDensities(/*regionIdx=*/0, rhoRefO, rhoRefG, rhoRefW);
     ccWaterPvt.setViscosity(/*regionIdx=*/0, 1);
```

**Comment by totto82** [comment]
@bska Thanks for the explanation. I will try it out tomorrow.

File: opm/core/simulator/EquilibrationHelpers.hpp (hunk 1 above)
Comment (andlaus) [reply opm/core/simulator/EquilibrationHelpers.hpp]: sure, but there is no real reason not to make `pc[BlackoilPhases::MaxNumPhases]` a local variable there?

Comment (andlaus) [reply opm/core/simulator/EquilibrationHelpers.hpp]: (same for the `fluidState`)

Comment (totto82) [reply opm/core/simulator/EquilibrationHelpers.hpp]: Good point. I will fix
